File: PadraoGoF/Criacional/FactoryMethod/FactoryMethod.js
// Aplicação do padrão GoF - Criacional - FactoryMethod

// Classe base de veiculos
class Veiculo {
    constructor(modelo) {
        this.modelo = modelo
    }

    mostrarDetalhes() {
        console.log(`Modelo: ${this.modelo}`)
    }
}

// Subclasse de veiculos
class Carro extends Veiculo {
    constructor(modelo) {
        super(modelo)
    }
}

class Moto extends Veiculo {
    constructor(modelo) {
        super(modelo)
    }
}

// Fabrica abstrata de veiculo
class FabricaDeVeiculos {
    criarVeiculo(modelo) {
        throw new Error('O método criarVeiculo deve ser implementado pelas subclasses')
    }
}

// Fabrica concreta de carros
class FabricaDeCarros extends FabricaDeVeiculos {
    criarVeiculo(modelo) {
        return new Carro(modelo)
    }
}

// Fabrica concreta de motos
class FabricaDeMotos extends FabricaDeVeiculos {
    criarVeiculo(modelo) {
        return new Moto(modelo)
    }
}

// Uso da Fabrica - Interface
const mercedes = new FabricaDeCarros()
const carro = mercedes.criarVeiculo('Sedan')
carro.mostrarDetalhes()

const fabricaMotos = new FabricaDeMotos()
const moto = fabricaMotos.criarVeiculo('Sport')
moto.mostrarDetalhes()

File: PadraoGoF/Criacional/FactoryMethod/FactoryMethod.py
# Aplicação do padrão GoF - Criacional - FactoryMethod

# Classe base de veiculos
class Veiculo:
    def __init__(self, modelo):
        self.modelo = modelo

    def mostrarDetalhes(self):
        print(f"Modelo: {self.modelo}")

# Subclasse de veiculos
class Carro (Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

class Moto (Veiculo):
    def __init__(self, modelo):
        super().__init__(modelo)

# Fabrica abstrata de veiculo
class FabricaDeVeiculos:
    def criarVeiculo(self, modelo):
        raise NotImplementedError('O método criarVeiculo deve ser implementado pelas subclasses')

# Fabrica concreta de carros
class FabricaDeCarros (FabricaDeVeiculos):
    def criarVeiculo(modelo):
        return Carro(modelo)

# Fabrica concreta de motos
class FabricaDeMotos (FabricaDeVeiculos):
    def criarVeiculo(modelo):
        return Moto(modelo)

# Uso da Fabrica - Interface
mercedes = FabricaDeCarros()
carro = mercedes.criarVeiculo('Sedan')
carro.mostrarDetalhes()

fabricaMotos = FabricaDeMotos()
moto = fabricaMotos.criarVeiculo('Sport')
moto.mostrarDetalhes()